Docs: add rust development manual (#555)
nervosnetwork/docs.nervos.org: website/docs/script/rust/rust-quick-start.mdx

- Using `ckb-debugger`
- Simulating an on-chain environment with `ckb-testtool`

This section covers the latter two methods: `ckb-debugger` and `ckb-testtool`.

#### Running with `ckb-debugger`:

```shell
ckb-debugger --bin build/release/hello-world
```

Output:

```text
Script log: Hello World!
Run result: 0
All cycles: 7366(7.2K)
```

#### Running Tests

Run the `test_hello_world` test ( Rust unit test):

```shell
cargo test -- tests::test_hello_world --nocapture
```

or

```shell
make test CARGO_ARGS="-- tests::test_hello_world --nocapture"
```

Output:

```text
---- tests::test_hello_world stdout ----
[contract debug] Hello World!
consume cycles: 7366
```

:::note
Formal contracts will get transaction information or on-chain data. For proper testing, it's best to script a complete transaction within tests. This will be covered in later chapters.
:::

## Simple Script (Print args data)

In real contracts, typically need to access transaction information or on-chain data. To improve upon the previous example, we introduce:

- `ckb-testtool`: Adding logic code to existing tests.
- `ckb-debugger`: Using `--tx-file` (or `-f`) to provide transaction data.

> The advantage of `ckb-testtool` is the flexibility to modify transaction details, making testing more convenient.

Reviewer: ```suggestion
**The advantage of `ckb-testtool`**:
- Allows modifying transaction details flexibly, making testing more convenient.
```

> Additionally, since `ckb-testtool` shares the same underlying code as `CKB`, it closely resembles the actual Nervos network.

Reviewer: ```suggestion
- `ckb-testtool` shares the same underlying code as `CKB`, closely resembling the actual Nervos Network
```

> Conversely, `ckb-debugger` allows direct command-line execution and supports debugging via `--mode gdb`. However, it requires manually constructing a transaction.

Reviewer: ```suggestion
**The advantage of `ckb-testtool`**:
- Allows direct command-line execution
- Supports debugging via `--mode gdb`
However, `ckb-debugger` requires manually constructing a transaction.
```

> A common workflow is to prioritize `ckb-testtool` for testing. If encountering complex bugs or other issues, transactions can be dumped and analyzed using ckb-debugger.

Reviewer: ```suggestion
**Recommended workflow**:
- Use `ckb-testtool` for regular testing
- If issues arise, use `ckb-debugger` to analyze transactions
```

> ### Code

Reviewer: ```suggestion
### Generate the Contract
```

            0
        }
        Err(err) => {
            ckb_std::debug!("load script failed: {:?}", err);
            -1
        }
    }
}
```

This code get args via `ckb_std::high_level::load_script` and prints them.

### Run With [ckb-testtool](https://crates.io/crates/ckb-testtool)

Executing `test_simple_print_args` will output:

```text
[contract debug] Args Len: 1
[contract debug] Args Data: [2a]
```

In the automatically generated code, args is set to `[42]`.

If we modify this section:

```rust
// prepare scripts
let lock_script = context
    .build_script(&out_point, Bytes::from(vec![42]))
    .expect("script");
```

to:

```rust
// prepare scripts
let args_data = ckb_testtool::context::random_hash().as_bytes();
let lock_script = context.build_script(&out_point, args_data).unwrap();
```

Output:

```text
[contract debug] Args Len: 32
[contract debug] Args Data: [3a, b5, fb, 71, 5f, 11, 7a, 54, cf, 90, 7f, cf, 5d, d8, 5c, 05, 5a, 31, 8d, b5, b2, 7e, 2e, 41, 90, 57, 96, cd, 0b, de, b2, 60]
```

Here, we use `ckb_testtool::context::random_hash()` to generate a 32-byte random value, which is passed as an argument to the contract.

### Run with ckb-debugger

> Since this contract requires transaction, executing it like before:

Reviewer: ```suggestion
Since this contract requires a transaction, executing it as before with:
```

    "version": "0x0",
    "cell_deps": [
      {
        "out_point": {
          "tx_hash": "0xbb0ff58103ed8a6dfd16480c0e736dbc469a5cf38d111778e29d91d0e01cd9da",
          "index": "0x0"
        },
        "dep_type": "code"
      }
    ],
    "header_deps": [],
    "inputs": [
      {
        "since": "0x0",
        "previous_output": {
          "tx_hash": "0x13f0197e4b72ad3c60e229dc0043661dbd7dc9e569a2a94672da6494e52241f7",
          "index": "0x0"
        }
      }
    ],
    "outputs": [
      {
        "capacity": "0x1f4",
        "lock": {
          "code_hash": "0x3bb06b94457b32e22b874951710c100258a480321e31c709cb0cb176edee4fcb",
          "hash_type": "type",
          "args": "0xefc0aa60c7af052a4aec89d0196905e77b10d07bbbeceef824f2d905ee5b71b4"
        },
        "type": null
      },
      {
        "capacity": "0x1f4",
        "lock": {
          "code_hash": "0x3bb06b94457b32e22b874951710c100258a480321e31c709cb0cb176edee4fcb",
          "hash_type": "type",
          "args": "0xefc0aa60c7af052a4aec89d0196905e77b10d07bbbeceef824f2d905ee5b71b4"
        },
        "type": null
      }
    ],
    "outputs_data": ["0x", "0x"],
    "witnesses": []
  }
}
```

:::note

> The `data` field contains the contract binary, which is too long to display here (this documentation), is replaced by `...`.

Reviewer: ```suggestion
The `data` field contains the contract binary, which is too long to display in this documentation; therefore, it is replaced by `...`.
```

:::

Save this JSON as a file and execute the contract with `ckb-debugger`:

```shell
ckb-debugger -f tests/test-vectors/test_simple_print_args.json
```

Output

```text
The cell_index is not specified. Assume --cell-index = 0
Script log: Args Len: 32
Script log: Args Data: [ef, c0, aa, 60, c7, af, 05, 2a, 4a, ec, 89, d0, 19, 69, 05, e7, 7b, 10, d0, 7b, bb, ec, ee, f8, 24, f2, d9, 05, ee, 5b, 71, b4]
Run result: 0
All cycles: 49239(48.1K)
```

To avoid manually updating the transaction file every time the contract is modified, use macros:

```
"data": "0x{{ data ../../build/release/simple-print-args }}",
```

This converts the compiled contract into data.

Similarly, replace the `Type Script` definition with:

```
"type": "{{ def_type simple-print-args }}"
```

and reference the `code_hash` using:

```
"output": {
    "capacity": "0x3e8",
    "lock": {
        "code_hash": "0x{{ ref_type simple-print-args }}",
        "hash_type": "type",
        "args": "0x11223344556677889900aabbccddeeff"
    },
    "type": null
},
```

This approach eliminates the need to update transaction details manually after each modification.

> The transaction file:

Reviewer: ```suggestion
With the updated transaction file below:
```

          "lock": {
            "code_hash": "0x0000000000000000000000000000000000000000000000000000000000000003",
            "hash_type": "data",
            "args": "0x"
          },
          "type": "{{ def_type simple-print-args }}"
        },
        "data": "0x{{ data ../../build/release/simple-print-args }}",
        "header": null
      }
    ],
    "header_deps": [],
    "extensions": []
  },
  "tx": {
    "version": "0x0",
    "cell_deps": [
      {
        "out_point": {
          "tx_hash": "0x0000000000000000000000000000000000000000000000000000000000000002",
          "index": "0x0"
        },
        "dep_type": "code"
      }
    ],
    "header_deps": [],
    "inputs": [
      {
        "since": "0x0",
        "previous_output": {
          "tx_hash": "0x0000000000000000000000000000000000000000000000000000000000000001",
          "index": "0x0"
        }
      }
    ],
    "outputs": [
      {
        "capacity": "0x1f4",
        "lock": {
          "code_hash": "0x{{ ref_type simple-print-args }}",
          "hash_type": "data2",
          "args": "0x"
        },
        "type": null
      }
    ],
    "outputs_data": ["0x"],
    "witnesses": []
  }
}
```

> Execute with ckb-debugger:

Reviewer: ```suggestion
execute with `ckb-debugger`:
```

```shell
ckb-debugger -f tests/test-vectors/test_simple_print_args.json
```

Output:

```
The cell_index is not specified. Assume --cell-index = 0
Script log: Args Len: 16
Script log: Args Data: [11, 22, 33, 44, 55, 66, 77, 88, 99, 00, aa, bb, cc, dd, ee, ff]
Run result: 0
All cycles: 38030(37.1K)
```

:::note
To improve the readability of this JSON, the following modifications have been made:

- The random value for `Args` has been changed to `0x11223344556677889900aabbccddeeff`.
- The random `tx_hash` has been replaced with a more structured value, such as `0x0000000000000000000000000000000000000000000000000000000000000002`.

These changes do not affect the actual execution results.
:::

## The End

> These two examples provide a simplified introduction to developing CKB scripts with Rust. Many details, such as additional `ckb-std` APIs and using gdb for debugging in `ckb-debugger`, have been omitted. These topics will be covered in later chapters.

Reviewer: ```suggestion
These two examples provide a simplified introduction to developing CKB Scripts with Rust. More advanced methods, such as additional `ckb-std` APIs and using `gdb` for debugging with `ckb-debugger`, are not covered here but will be explored in future chapters.
```

---
id: rust-quick-start
title: "Quick Start"
---

import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";
import Link from "@docusaurus/Link";
import Tooltip from "@components/Tooltip";
import TutorialHeader from "@components/TutorialHeader";
import ScriptTools from "./../_ScriptTools.mdx";

# Quick Start

<TutorialHeader
  time={" 5 - 15min "}
  tools={"script"}
  customTools={[
    <div key="debugger">
      <Link
        to="/docs/getting-started/installation-guide#5-ckb-debugger-01170"
        target="_blank"
        rel="noopener noreferrer"
      >
        CKB-Debugger (≥0.117.0)
      </Link>
    </div>,
  ]}
/>

This chapter provides a step-by-step guide on developing contracts with Rust on CKB with several examples. As a "Quick Start" guide, it will not diving into in-depth details. For a more comprehensive understanding, please refer to the subsequent chapters.

> :::note

Reviewer 2: please remove this note. too many notes here.

Before proceeding, you should be familiar with:

- CKB basics and the transaction structures
- Rust
- The `make` command
  :::

> ## Why Rust for CKB Contracts?

Reviewer 2: remove this subtitle

> Rust is widely used in developing core components of CKB due to its high performance and ease of use. Scripts often involve operations like signing, hashing, and handling large numbers. Rust’s superior performance helps reduce <Tooltip>Cycles</Tooltip> consumption. Furthermore, as Scripts are critical for asset management, Rust’s memory safety features help minimize vulnerabilities, making it a **recommended choice for Script development** .

Reviewer 2: let's put this why-rust-for-ckb-contracts explaination to the https://docs.nervos.org/docs/script/program-language-for-script 's rust section.

> This chapter will demonstrate the complete process of developing CKB Scripts with Rust using two examples. Key steps include:

Reviewer 2: please merge this into the above introduction.
